### ISSUE TITLE
Won't converge for simple data

### ISSUE THREAD

**Issue by shamoons** [cazala/synaptic#14]
```
keyLength = 30;
inputLayer = new Layer(keyLength);

hiddenLayers = [new Layer(keyLength), new Layer(parseInt(keyLength / 2))];

outputLayer = new Layer(1);

inputLayer.project(hiddenLayers[0]);

hiddenLayers[0].project(hiddenLayers[1]);

hiddenLayers[1].project(outputLayer);

trainedNetwork = new Network({
  input: inputLayer,
  hidden: hiddenLayers,
  output: outputLayer
});

iterations = 0;

MSE = 1;

learningRate = 0.3;

while (!(iterations > 20000 || MSE < 0.000005)) {
  loopErrorTotal = 0;
  _.each(trainingData, function(trainingSet) {
    var currentOutput;
    currentOutput = void 0;
    currentOutput = trainedNetwork.activate(trainingSet.input);
    trainedNetwork.propagate(learningRate, [trainingSet.output]);
    return loopErrorTotal += Math.pow(currentOutput[0] - trainingSet.output, 2);
  });
  iterations++;
  MSE = loopErrorTotal / trainingData.length;
  console.log('Iteration: ' + iterations + '\u9MSE: ' + MSE);
}
```

That's my code. My `trainingData` only has 2 samples in it. I want to just get some basic convergence to get my error management correct. However, my `MSE` steadily increases instead of decreases. Not sure what I'm doing wrong. Any thoughts?

**Comment by cazala** [comment]
there's nothing wrong with that code, but the error sometimes goes up in the beginning (specially when you have many connections), or sometimes doesn't go up but its stuck for a few thousand iterations until it escapes the local minima and then it goes down really fast. I tested your code using this training set:

```
trainingData = [
{ input:[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0], output: 0},
{ input:[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0], output: 1}
]
```

it reached the 20k iterations and the MSE was 0.802..., then I changed the iterations limit to 200k and it converged at Iteration: 153000, MSE: 0.000008...

```
Iteration: 147000 MSE: 0.8021047553849563
Iteration: 148000 MSE: 0.8020910343167962
Iteration: 149000 MSE: 0.802050928980153
Iteration: 150000 MSE: 0.8014304108388679
Iteration: 151000 MSE: 0.00042315614372781865
Iteration: 152000 MSE: 0.0000253250799368847
Iteration: 153000 MSE: 0.000008267952733910032
```

also before you test that, replace the console.log line with this:

```
if (iterations%1000==0)
  console.log('Iteration: ' + iterations + ' MSE: ' + MSE);
```

So you don't print every single iteration (it slows down the execution a lot). And you can play tweaking the learning rate to achieve convergence faster.

**Comment by cazala** [comment]
btw, you can achieve the same result as that code by using a trainer, like this:

```
var trainer = new Trainer(trainedNetwork);
trainer.train(trainingData, {
    rate: .1,
    iterations: 200000,
    error: 0.000005,
    log: 1000,
    cost: Trainer.cost.MSE
});
```

the trainer includes also an alternative cost function: `Trainer.cost.CROSS_ENTROPY`
